subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/LockOnDemandCrossProcessBuildCacheService.java

```diff
@@ -120,13 +126,22 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) throws Buil
         return crossProcessCacheAccess.withFileLock(() -> delegate.load(key, reader));
     }
 
+    @Override
+    public void cleanup() {
+        crossProcessCacheAccess.withFileLock(() -> {
+            delegate.cleanup();
+            return null;
+        });
+    }
+
     @Override
     public synchronized void close() {
         withOwnershipNow(() -> {
             try {
                 if (fileLockHeldByOwner != null) {
                     fileLockHeldByOwner.run();
                 }
+                persistentCache.close();
                 crossProcessCacheAccess.close();
```

Comment:
:x: Let's use a `Closer` here to make sure both of these get closed even if the first throws an exception.